server/base/src/main/java/org/apache/accumulo/server/util/serviceStatus/ServiceStatusReport.java

```diff
@@ -0,0 +1,168 @@
+        .reduce(Integer::sum).orElse(0);
+    this.noHosts = noHosts;
+    this.summaryMap = summaryMap;
+  }
+
+  public String getReportTime() {
+    return reportTime;
+  }
+
+  public Map<ReportKey,StatusSummary> getSummaryMap() {
+    return summaryMap;
+  }
+
+  public String toJson() {
+    return gson.toJson(this, ServiceStatusReport.class);
+  }
+
+  public String report(final StringBuilder sb) {
+    sb.append("Report time: ").append(rptTimeFmt.format(ZonedDateTime.now(ZoneId.of("UTC"))))
+        .append("\n");
+    int zkErrors = summaryMap.values().stream().map(StatusSummary::getErrorCount)
+        .reduce(Integer::sum).orElse(0);
+    sb.append("ZooKeeper read errors: ").append(zkErrors).append("\n");
+
+    writeServiceStatus(sb, ReportKey.MANAGER, summaryMap.get(ReportKey.MANAGER), noHosts);
+    writeServiceStatus(sb, ReportKey.MONITOR, summaryMap.get(ReportKey.MONITOR), noHosts);
+    writeServiceStatus(sb, ReportKey.GC, summaryMap.get(ReportKey.GC), noHosts);
+    writeServiceStatus(sb, ReportKey.T_SERVER, summaryMap.get(ReportKey.T_SERVER), noHosts);
+    writeResourceGroups(sb, ReportKey.S_SERVER, summaryMap.get(ReportKey.S_SERVER), noHosts);
+    writeServiceStatus(sb, ReportKey.COORDINATOR, summaryMap.get(ReportKey.COORDINATOR), noHosts);
+    writeResourceGroups(sb, ReportKey.COMPACTOR, summaryMap.get(ReportKey.COMPACTOR), noHosts);
+
+    sb.append("\n");
+    LOG.trace("fmtStatus - with hosts: {}", summaryMap);
+    return sb.toString();
+  }
+
+  public void writeServiceStatus(final StringBuilder sb, final ReportKey displayNames,
+      final StatusSummary summary, boolean noHosts) {
+    if (summary == null) {
+      sb.append(displayNames).append(": unavailable").append("\n");
+      return;
+    }
+
+    writeCounts(sb, summary);
+
+    // skip host info if requested
+    if (noHosts) {
+      return;
+    }
+    if (summary.getServiceCount() > 0) {
+      var hosts = summary.getServiceNames();
+      hosts.forEach(h -> sb.append(I2).append(h).append("\n"));
+    }
+  }
+
+  private void writeCounts(StringBuilder sb, StatusSummary summary) {
+    sb.append(summary.getDisplayName()).append(": count: ").append(summary.getServiceCount());
+    if (summary.getErrorCount() > 0) {
+      sb.append(", (ZooKeeper errors: ").append(summary.getErrorCount()).append(")\n");
+    } else {
+      sb.append("\n");
+    }
+  }
+
+  private void writeResourceGroups(final StringBuilder sb, final ReportKey reportKey,
+      final StatusSummary summary, boolean noHosts) {
+    if (summary == null) {
+      sb.append(reportKey).append(": unavailable").append("\n");
+      return;
+    }
+
+    writeCounts(sb, summary);
+
+    // skip host info if requested
+    if (noHosts) {
+      return;
+    }
+    if (!summary.getResourceGroups().isEmpty()) {
+      sb.append(I2).append("resource groups:\n");
+      summary.getResourceGroups().forEach(g -> sb.append(I4).append(g).append("\n"));
```

Comment:
Do we need to print the groups twice? I wonder if we could print something like:
```
  Group: DEFAULT
    localhost:1234
    localhost:1235
  Group: Group1
    localhost:1236
```

Reply:
Address in f5f92218bd